src/runtime/swg-client-event-manager.js

```diff
@@ -0,0 +1,312 @@
+   * @param {Object} value
+   */
+  setAdditionalParameters(value) {
+    assert(SwgClientEvent.isValidAdditionalParameters_(value),
+        'Invalid additionalParameters in SwgClientEvent (' + value + ')');
+    this.additionalParameters_ = value;
+  }
+
+  /**Returns true if the object is valid.  Otherwise it throws an error if you
+   * pass true and returns false if you don't.
+   * @param {boolean} createError
+   */
+  isValid(createError) {
+    if (!SwgClientEvent.isValidType_(this.eventType_)) {
+      if (!createError) {
+        return false;
+      }
+      throw new Error('Invalid event type in SwgClientEvent (' +
+          this.eventType_ + ')');
+    }
+    if (!SwgClientEvent.isValidOriginator_(this.eventOriginator_)) {
+      if (!createError) {
+        return false;
+      }
+      throw new Error('Invalid event originator in SwgClientEvent (' +
+          this.eventOriginator_ + ')');
+    }
+    if (!SwgClientEvent.isValidIsFromUserAction_(this.isFromUserAction_)) {
+      if (!createError) {
+        return false;
+      }
+      throw new Error('Invalid isFromUserAction in SwgClientEvent (' +
+          this.isFromUserAction_ + ')');
+    }
+    if (!SwgClientEvent.isValidAdditionalParameters_(
+        this.additionalParameters_)) {
+      if (!createError) {
+        return false;
+      }
+      throw new Error('Invalid additional parameters in SwgClientEvent (' +
+          this.additionalParameters_ + ')');
+    }
+    return true;
+  }
+}
+
+export class SwgClientEventManager {
+
+  /**True if the value is a function
+   * @param {*} value
+   * @private
+   */
+  static isFunction_(value) {
+    return typeof value === 'function';
+  }
+
+  constructor() {
+    /** @type {!CallBackArray} @private */
+    this.listeners_ = [];
+
+    /** @type {!CallBackArray} @private */
+    this.filterers_ = [];
+  }
+
+  /**Remove all existing filterers and listeners
+   */
+  clear() {
+    this.listeners_ = [];
+    this.filterers_ = [];
+  }
+
+  /**
+   * Ensures the callback function is notified anytime one of the passed
+   * events occurs unless a filterer returns false.
+   * @param {!SwgEventManagerListener} callback
+   * @public
+   */
+  addListener(callback) {
+    assert(SwgClientEventManager.isFunction_(callback),
+        'Invalid callback in SwgClientEventManager.addListener');
+    this.listeners_.push(callback);
+  }
+
+  /**
+   * Filterers are called before listeners, if it returns
+   * ShouldFilter.STOP_EXECUTING then the event is canceled.
+   * @param {!SwgEventManagerFilterer} callback
+   * @public
+   */
+  addFilterer(callback) {
+    assert(SwgClientEventManager.isFunction_(callback),
+        'Invalid callback in SwgClientEventManager.addFilterer');
+    this.filterers_.push(callback);
+  }
+
+  /**Call this function to inform all listeners that a client event has
+   * occurred.
+   * @param {!SwgClientEvent} event
+   * @public
+   */
+  logEvent(event) {
```

Comment:
Since we are directly invoking callbacks within this method, we need to make sure the call to logEvent is non blocking. Can we think of ways to do so? Again please feel to refer me to this design consideration in a doc if you have already addressed it.

Reply:
I believe this is done